src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -2030,16 +2031,39 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 	workerExtendedOpNode->hasDistinctOn = false;
 	workerExtendedOpNode->hasWindowFuncs = originalOpNode->hasWindowFuncs;
 	workerExtendedOpNode->windowClause = originalOpNode->windowClause;
+	workerExtendedOpNode->groupClauseList = groupClauseList;
+
+	/*
+	 * We are only pushing down distinct on, and limit (along with order by)
+	 * if distinct on clause is superset of group by clause. If there is no
+	 * grouping we are good to pushdown distinct (and orderby/limit). Query
+	 * result depends on order by expression.
+	 */
+	if (shouldPushdownOrderByLimit && originalOpNode->distinctClause &&
```

Comment:
Well, I think I found what disturbs me here.

We're creating a cyclic dependency among `limit pushdown` and `distinct/distinct on`. We initially look `shouldPushdownOrderByLimit ` and `DISTINCT` to decide on the new value of `shouldPushdownOrderByLimit `. Later, we use `shouldPushdownOrderByLimit ` to decide on `shouldPushdownDistinct `. 

What about if we break the dependency and go only in one direction that we always evaluate `DISTINCT/DISTINCT ON` and modify `shouldPushdownOrderByLimit` whenever necessary. 

Such as this: 

https://github.com/citusdata/citus/compare/fix_distinct_on_agg...fix_distinct_on_agg_onder?expand=1

(Note that the branch is NOT complete, I mostly wanted to share my take on the issue. We should probably iterate on the correctness and comments etc once more)